Author: Tian Shilin <tianshilin@uniontech.com>
Date:   Wed Dec 18 10:10:08 2024
Subject: QTextCodec: avoid read-past-buffer in codecForName()
Upstream: https://codereview.qt-project.org/c/qt/qt5compat/+/723911


---
Index: qt6-5compat/src/core5/codecs/qtextcodec.cpp
===================================================================
--- qt6-5compat.orig/src/core5/codecs/qtextcodec.cpp
+++ qt6-5compat/src/core5/codecs/qtextcodec.cpp
@@ -48,6 +48,7 @@
 #endif // icu
 
 #include <mutex>
+#include <string>
 
 #include <stdlib.h>
 #include <ctype.h>
@@ -508,6 +509,10 @@ QTextCodec *QTextCodec::codecForName(con
     if (name.isEmpty())
         return nullptr;
 
+    // ensure NUL-termination:
+    const std::string name0(name.data(), size_t(name.size()));
+    // (do it outside the critical section, even if we may not need it)
+
     const TextCodecsMutexLocker locker;
 
     QTextCodecData *globalData = QTextCodecData::instance();
@@ -524,14 +529,14 @@ QTextCodec *QTextCodec::codecForName(con
 
     for (TextCodecListConstIt it = globalData->allCodecs.constBegin(), cend = globalData->allCodecs.constEnd(); it != cend; ++it) {
         QTextCodec *cursor = *it;
-        if (qTextCodecNameMatch(cursor->name(), name)) {
+        if (qTextCodecNameMatch(cursor->name().constData(), name0.data())) {
             if (cache)
                 cache->insert(name, cursor);
             return cursor;
         }
         QList<QByteArray> aliases = cursor->aliases();
         for (ByteArrayListConstIt ait = aliases.constBegin(), acend = aliases.constEnd(); ait != acend; ++ait) {
-            if (qTextCodecNameMatch(*ait, name)) {
+            if (qTextCodecNameMatch(ait->constData(), name0.data())) {
                 cache->insert(name, cursor);
                 return cursor;
             }
@@ -540,7 +545,7 @@ QTextCodec *QTextCodec::codecForName(con
 
     return nullptr;
 #else
-    return QIcuCodec::codecForNameUnlocked(name);
+    return QIcuCodec::codecForNameUnlocked(name0.data());
 #endif
 }
 
Index: qt6-5compat/tests/auto/core5/codecs/qtextcodec/tst_qtextcodec.cpp
===================================================================
--- qt6-5compat.orig/tests/auto/core5/codecs/qtextcodec/tst_qtextcodec.cpp
+++ qt6-5compat/tests/auto/core5/codecs/qtextcodec/tst_qtextcodec.cpp
@@ -2335,6 +2335,15 @@ void tst_QTextCodec::invalidNames()
     huge = huge + huge + huge + huge + huge + huge + huge + huge;
     huge = huge + huge + huge + huge + huge + huge + huge + huge;
     QVERIFY(!QTextCodec::codecForName(huge));
+
+    {
+        const char ISO_8859_15[] = "ISO-8859-15";
+        const QByteArray ISO_8859_1 = QByteArray::fromRawData(ISO_8859_15,
+                                                              sizeof ISO_8859_15 - 2);
+        const auto codec = QTextCodec::codecForName(ISO_8859_1);
+        QVERIFY(codec);
+        QCOMPARE(codec->name(), "ISO-8859-1");
+    }
 }
 
 void tst_QTextCodec::checkAliases_data()
